src/evd/cyclic/evd_cost.cpp

```diff
@@ -172,25 +171,24 @@ size_t blocked_cost_with_subprocedure_evd(size_t n, size_t b, size_t n_iter, siz
     size_t single_mult_block_mul = b * b * b;
     size_t single_add = b * b;
 
-    size_t block_adds = loops * (4 * n_blocks * (12 * single_mult_block_add + 6 * single_add));
-    size_t block_muls = loops * (4 * n_blocks * (12 * single_mult_block_mul));
-    size_t individual_block_ops = loops * subprocedure_cost(4 * b, individual_block_iter);  // 4 individual blocks
+    size_t block_adds = loops * (3 * n_blocks * (2 * single_add + 4 * single_mult_block_add));
```

Comment:
Ok, now let me explain how I designed the cost_function. See the corresponding code below:

```
for (int ep = 1; ep <= epoch; ep++) {
        for (size_t i_block = 0; i_block < n_blocks - 1; ++i_block) {
            for (size_t j_block = i_block + 1; j_block < n_blocks; ++j_block) {
                copy_block(Amat, i_block, i_block, Ablockmat, 0, 0, block_size);
                copy_block(Amat, i_block, j_block, Ablockmat, 0, 1, block_size);
                copy_block(Amat, j_block, i_block, Ablockmat, 1, 0, block_size);
                copy_block(Amat, j_block, j_block, Ablockmat, 1, 1, block_size);

                // evd_block_vector(Ablockmat, Vblockmat);
                evd_subprocedure_vectorized(Ablockmat, Vblockmat);

                matrix_transpose(Vblockmat, Vblockmat);

                for (size_t k_block = 0; k_block < n_blocks; ++k_block) {
                    mult_block(Vblockmat, 0, 0, Amat, i_block, k_block, M1mat, 0, 0, block_size);
                    mult_block(Vblockmat, 0, 1, Amat, j_block, k_block, M2mat, 0, 0, block_size);
                    matrix_add(M1mat, M2mat, M2mat);
                    mult_block(Vblockmat, 1, 0, Amat, i_block, k_block, M1mat, 0, 0, block_size);
                    copy_block(M2mat, 0, 0, Amat, i_block, k_block, block_size);
                    mult_block(Vblockmat, 1, 1, Amat, j_block, k_block, M2mat, 0, 0, block_size);
                    matrix_add(M1mat, M2mat, M2mat);
                    copy_block(M2mat, 0, 0, Amat, j_block, k_block, block_size);
                }
                matrix_transpose(Vblockmat, Vblockmat);

                for (size_t k_block = 0; k_block < n_blocks; ++k_block) {
                    mult_block(Amat, k_block, i_block, Vblockmat, 0, 0, M1mat, 0, 0, block_size);
                    mult_block(Amat, k_block, j_block, Vblockmat, 1, 0, M2mat, 0, 0, block_size);
                    matrix_add(M1mat, M2mat, M2mat);
                    mult_block(Amat, k_block, i_block, Vblockmat, 0, 1, M1mat, 0, 0, block_size);
                    copy_block(M2mat, 0, 0, Amat, k_block, i_block, block_size);
                    mult_block(Amat, k_block, j_block, Vblockmat, 1, 1, M2mat, 0, 0, block_size);
                    matrix_add(M1mat, M2mat, M2mat);
                    copy_block(M2mat, 0, 0, Amat, k_block, j_block, block_size);
                }

                for (size_t k_block = 0; k_block < n_blocks; ++k_block) {
                    mult_block(Eigen_vectors, k_block, i_block, Vblockmat, 0, 0, M1mat, 0, 0, block_size);
                    mult_block(Eigen_vectors, k_block, j_block, Vblockmat, 1, 0, M2mat, 0, 0, block_size);
                    matrix_add(M1mat, M2mat, M2mat);
                    mult_block(Eigen_vectors, k_block, i_block, Vblockmat, 0, 1, M1mat, 0, 0, block_size);
                    copy_block(M2mat, 0, 0, Eigen_vectors, k_block, i_block, block_size);
                    mult_block(Eigen_vectors, k_block, j_block, Vblockmat, 1, 1, M2mat, 0, 0, block_size);
                    matrix_add(M1mat, M2mat, M2mat);
                    copy_block(M2mat, 0, 0, Eigen_vectors, k_block, j_block, block_size);
                }
            }
        }
    }
```
1. The block size of Vblockmat and Ablockmat is 2b * 2b = 4 * (b*b)
2. `mult_block` is executed 12 times in the inner loop => both block level adds and mults are performed 12 * (single block = b*b*b) times
So, from 1 & 2, the overall count is (4 * n_blocks * (12 * single_mult_block_mul))

Reply:
OK, I see 3 loops, hence `3*n_blocks`. In each of these I see `4 mult block` which each computes `b*b*b adds` and `b*b*b` muls.  I do not see that they execute `4*b*b*b` muls/adds. In the defintion of `mult_block` i cannot see the loop you are talking about.